homeworks/08-js-advanced-1/script.js

```diff
@@ -0,0 +1,216 @@
+
+      hour: dateString.substring(format.indexOf("h"), format.indexOf("h") + 2),
+
+      min: dateString.substring(format.indexOf("m"), format.indexOf("m") + 2),
+
+      sec: dateString.substring(format.indexOf("s"), format.indexOf("s") + 2),
+    };
+    return new Date(...Object.values(date).map((item) => parseInt(item)));
+  }
+  // E.g. Today is 23.02.2021, 14:00:00
+  console.log(offset(moment("23/02/2021 13:30:00", "DD/MM/YYYY hh:mm:ss")));
+  // 30 minutes ago
+
+  console.log(offset(moment("23/02/2021 13:00:00", "DD/MM/YYYY hh:mm:ss")));
+  // 1 hour ago
+
+  console.log(offset(moment("23/02/2021 11:30:00", "DD/MM/YYYY hh:mm:ss")));
+  // 2 hours 30 minutes ago
+
+  console.log(offset(moment("22/02/2021 14:00:00", "DD/MM/YYYY hh:mm:ss")));
+  // 1 day ago
+
+  console.log(offset(moment("23/02/2020 10:00:00", "DD/MM/YYYY hh:mm:ss")));
+  // 366 days ago
+})();
+
+// 4. Random dates
+// Create a function that generate a random date between to dates
+import mm from "moment";
+(() => {
+  function randomDate(date1, date2) {
+    const diff = date2.valueOf() - date1.valueOf();
+    const randomDiff = Math.random() * diff;
+    const randomTimestamp = date1.valueOf() + randomDiff;
+    return mm(randomTimestamp);
+  }
+  const date1 = mm("23/01/2021", "DD/MM/YYYY");
+  const date2 = mm("23/02/2021", "DD/MM/YYYY");
+  console.log(randomDate(date1, date2).format("DD/MM/YY"));
+  // 20/02/2021
+})();
+
+//https://www.codewars.com/kata/merged-objects
+function objConcat(arr) {
+  let res = {};
+  for (let obj of arr) {
+    for (let key in obj) {
+      res[key] = obj[key];
+    }
+  }
+  return res;
+}
+//https://www.codewars.com/kata/547f1a8d4a437abdf800055c/train/javascript
+function NamedOne(first, last) {
+  this.firstName = first;
+  this.lastName = last;
+  Object.defineProperty(this, "fullName", {
+    set: function (str) {
+      const names = str.split(" ");
+      if (names.length == 2) {
+        this.firstName = names[0];
+        this.lastName = names[1];
+      }
+    },
+    get: function () {
+      return this.firstName + " " + this.lastName;
+    },
+  });
+}
+//https://www.codewars.com/kata/54834b3559e638b39d0009a2/train/javascript
+function OnceNamedOne(first, last) {
+  Object.defineProperty(this, "firstName", {
+    value: first,
+  });
+
+  Object.defineProperty(this, "lastName", {
+    value: last,
+  });
+
+  Object.defineProperty(this, "fullName", {
+    value: this.firstName + " " + this.lastName,
+  });
+}
+//https://www.codewars.com/kata/partial-keys
+function partialKeys(obj) {
+  const newObj = {};
+  function generateKeys(key) {
+    const partials = [];
+    for (let i = 1; i <= key.length; i++) {
+      partials.push(key.slice(0, i));
+    }
+    return partials;
+  }
+  for (let key of Object.keys(obj).sort()) {
+    const partials = generateKeys(key);
+    for (let partialKey of partials) {
+      if (!(partialKey in newObj)) {
+        newObj[partialKey] = obj[key];
+      }
+    }
+  }
+  return newObj;
+}
```

Comment:
<img width="1131" alt="Screenshot 2024-05-18 at 17 51 47" src="https://github.com/mg300/solvd/assets/97586975/d59e0009-560b-4190-b771-2ab1255d76fa">

It did not pass tests